---
title: HTtools
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
params:
        fn: '../test/config/config_SN.yaml'
        location: NULL
---

Last run: `r date()`

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
cache.opt <- TRUE
eval.opt <- ifelse(is.null(params$location), TRUE, FALSE)
```

```{r imports, results='asis'}
library(DT)
library(stringr)
library(plotly)
library(ggplot2)
library(plyr)
library(dplyr)
library(tidyr)
library(yaml)
library(heatmaply)
library(purrr)
library(RColorBrewer)
```

```{r, results='asis'}
subchunkify <- function(g) {
  g_deparsed <- paste0(deparse(
    function() {g}
  ), collapse = '')

  sub_chunk <- paste0("
  `","``{r sub_chunk_", floor(runif(1) * 10000), ", fig.height=10, echo=FALSE}",
  "\n(", 
    g_deparsed
    , ")()",
  "\n`","``
  ")

  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
  }
```

```{r data_setup}
config <- read_yaml(params$fn)
mydir <- paste0('../data/', config[['name']])
dir.create(paste0(mydir, '/ORFmap'), showWarnings=FALSE)
dir.create(paste0(mydir, '/heatmap'), showWarnings=FALSE)
logdir <- paste0(mydir, '/logs/')
logs <- list()
logs[['fastqscreen']] <- paste0(logdir, 'fastq_screen_', config[['name']], '_log.txt')
logs[['integration']] <- paste0(logdir, 'integration_', config[['name']], '_log.txt')
logs[['location']] <- paste0(logdir, 'location_', config[['name']], '_log.txt')

sample.list <- list()
for (sample in names(config[['sample']])){
    version <- config[['genomevs']][[config[['genome']]]]
    sample.list[[sample]][['screen']] <- paste0(mydir, '/fastqscreen/screen_', sample, '.fa')
    sample.list[[sample]][['screen_SN']] <- paste0(mydir, '/fastqscreen/screen_', sample, '_SN.txt')
    sample.list[[sample]][['screen_ids']] <- paste0(mydir, '/fastqscreen/screen_', sample, '_ids.txt')
    sample.list[[sample]][['screen_PBS']] <- paste0(mydir, '/fastqscreen/screen_PBS_', sample, '.fa')
    sample.list[[sample]][['screen_PBS_SN']] <- paste0(mydir, '/fastqscreen/screen_PBS_', sample, '_SN.txt')
    sample.list[[sample]][['screen_PBS_ids']] <- paste0(mydir, '/fastqscreen/screen_PBS_', sample, '_ids.txt')
    sample.list[[sample]][['blast']] <- paste0(mydir, '/blast/blast_', sample, '.', version, '.txt')
    sample.list[[sample]][['preintegration']] <- paste0(mydir, '/filblast/integration_', sample, '.', version, '.txt')
    sample.list[[sample]][['preintegration_multimatch']] <- paste0(mydir, '/filblast/integration_multimatch_',
                                                                   sample, '.', version, '.txt')
    sample.list[[sample]][['integration']] <- paste0(mydir, '/location/true_integration_', sample, '.', version, '.txt')
    sample.list[[sample]][['location']] <- paste0(mydir, '/location/location_', sample, '.', version, '.txt')
    sample.list[[sample]][['intergenic']] <- paste0(mydir, '/location/intergenic_', sample, '.', version, '.txt')
    sample.list[[sample]][['ORF']] <- paste0(mydir, '/location/ORF_', sample, '.', version, '.txt')
    sample.list[[sample]][['ltr3']] <- paste0(mydir, '/location/homologous_recombination/ltr3_integration_',
                                              sample, '.', version, '.txt')
    sample.list[[sample]][['ltr5']] <- paste0(mydir, '/location/homologous_recombination/ltr5_integration_',
                                              sample, '.', version, '.txt')
    sample.list[[sample]][['sololtr']] <- paste0(mydir, '/location/homologous_recombination/sololtr_integration_',
                                              sample, '.', version, '.txt')
    sample.list[[sample]][['exclude']] <- paste0(mydir, '/location/excluded/excluded_integration_',
                                              sample, '.', version, '.txt')
    logs[['blast']][[sample]] <- paste0(logdir, 'blast_', config[['name']], '_', sample, '.', version, '.log.txt')
}
```


```{r loc_setup, eval=!eval.opt}
fn <- basename(params$location)
sample.list <- list()
sample.list[[fn]][['location']] <- params$location
sample.list[[fn]][['intergenic']] <- str_replace(sample.list[[fn]][['location']] ,
                                                              'location/location',
                                                              'location/intergenic')
sample.list[[fn]][['ORF']] <- str_replace(sample.list[[fn]][['location']] ,
                                                              'location/location',
                                                              'location/ORF')
```


# Samples and parameters used {.tabset}

## Sampletable

```{r, results='asis', eval=!eval.opt}
cat('\n\nDisabled when location file provided in config\n\n')
```

```{r, results='asis', eval=eval.opt}
sampletable <- data.frame(config[['sample']][[1]])
if (length(names(config[['sample']])) > 1) {
    for (name in names(config[['sample']])[2:length(config[['sample']])]) {
        sampletable <- rbind.fill(sampletable,
                            data.frame(config[['sample']][[name]]))
    }
}
row.names(sampletable) <- names(config[['sample']])
knitr::kable(t(sampletable))
```

## List of fastq(s)

```{r, results='asis', eval=!eval.opt}
cat('\n\nDisabled when location file provided in config\n\n')
```

```{r, results='asis', eval=eval.opt}
for (fq in config[['fastq']]) {
    cat(paste0('\n- ', fq, '\n'))
}
```

## Sample-specific parameters {.tabset}

```{r, results='asis', eval=!eval.opt}
cat('\n\nDisabled when location file provided in config\n\n')
```

```{r, results='asis', eval=eval.opt}
chro_list <- config[[config[['chro_listvs']][[config[['genome']]]]]]
for (sample in names(config[['sample']])){
    cat(paste0('\n\n### ', sample, '\n\n'))
    subcfg <- config[['sample']][[sample]]
    scfg <- list(
        'genome'=config[['genome']],
        'genomevs'=config[['genomevs']][[config[['genome']]]],
        'genomedb'=config[['genomedb']][[config[['genome']]]],
        'chro_list'=paste(chro_list, collapse=', '),
        'gneomecds'=config[['genomecds']][[config[['genome']]]],
        "preexist_ltr.ltr5"=config[['preexist_ltr']][[subcfg[['lib_design']]]][['ltr5']],
        "preexist_ltr.ltr3"= config[['preexist_ltr']][[subcfg[['lib_design']]]][['ltr3']],
        "preexist_ltr.solo.ltr"=config[['preexist_ltr']][[subcfg[['lib_design']]]][['sololtr']],
        'ltrcircle'=config[['ltrcircle']][[subcfg[['lib_design']]]],
        'plasmid'=config[['plasmid']][[subcfg[['lib_design']]]],
        'primary_re'=config[['primary_re']][[subcfg[['lib_design']]]],
        'primary_incomplete'=config[['primary_incomplete']][[subcfg[['lib_design']]]],
        'second_re'=config[['second_re']][[subcfg[['lib_design']]]],
        'second_incomplete'=config[['second_incomplete']][[subcfg[['lib_design']]]],
        'dist_to_second_incomplete'=config[['dist_to_second_incomplete']][[subcfg[['lib_design']]]],
        'pbs'=config[['pbs']][[subcfg[['lib_design']]]],
        'tsd'=config[['tsd']][[subcfg[['integrase']]]],
        'exclude'=paste(config[['exclude']], collapse=', ')
    )
    print(knitr::kable(t(data.frame(scfg)), row.names=TRUE, col.names=c(sample)))
}
```

## HTtools parameters

Note: `legacy_mode` is whether output file formats match the original HTtools.

```{r, results='asis'}
hcfg <- list(
           'legacy_mode'=config[['legacy_mode']],
           'length_to_match'=config[['length_to_match']],
           'min_length'=config[['min_length']],
           'allowed_mismatches'=config[['allowed_mismatches']],
           'linker'=config[['linker']],
           'blastview'=config[['blastview']],
           'blastevalue'=config[['blastevalue']],
           'max_score_diff'=config[['max_score_diff']],
           'orf_map_interval'=config[['orf_map_interval']],
           'orf_map_window'=config[['orf_map_window']],
           'avg_orf_length'=config[['avg_orf_length']]
           )
knitr::kable(t(data.frame(hcfg)), row.names=TRUE, col.names=c('Parameters'))
```


# Filtering steps {.tabset}

```{r, results='asis', eval=!eval.opt}
cat('\n\nDisabled when location file provided in config\n\n')
```

## Fastq screen

```{r funcs, results='asis'}
barplot <- function(log, x, y='value', fill, filllab) {
    bp <- ggplot(log, aes_string(x=x, y=y, fill=fill)) +
    geom_bar(stat="identity", position=position_dodge()) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          axis.title = element_blank())
    labs(fill=filllab)
    return(bp)
}

bply <- function(bp, xlab, ylab= 'Number of sequence reads') {
    b <- ggplotly(bp) %>% layout(xaxis = list(title=xlab), yaxis = list(title=ylab))
    return(b)
}

catlink <- function(file, sample, legend) {
    file <- str_replace(file,
                        paste0('data/', config['name'], '/'),
                        '')
    cat(paste0('\n\ndata/', config['name'], '/\n\n'))
    cat(paste0(
        '\n- [', file, '](', file, ') : ', legend, ' file for ', sample ,' sample(s)\n'))
}
```


```{r fastqscreen, results='asis', eval=eval.opt}
loghead <- read.csv(logs[['fastqscreen']], sep='\t', nrows=2, header=FALSE)
cat(paste0(loghead[1,], '\n\n'))
log <- read.csv(logs[['fastqscreen']], sep='\t', skip=2) %>%
    # different columns are present whether the design has SN or not
    select(-one_of('Barcode.start', 'Barcode.length', 
                   'SN.start', 'SN.length',
                   'Transposon.end', 'Beginning.seq')) %>%
    mutate_at(vars(-Name),as.character) %>%
    mutate_at(vars(-Name),as.numeric) %>%
    pivot_longer(-Name, values_drop_na = TRUE) %>%
    mutate(name = factor(name, levels=unique(name)))
log <- log %>%
    dplyr::group_by(Name) %>%
    dplyr::mutate(percent = value / sum(value, na.rm=TRUE))
cat('\n\n### Raw reads\n\n')
bply(barplot(log=log, x='name', fill='Name', filllab='Sample'),
    xlab='Fastqscreen Filtering steps')
cat('\n\n### Percent reads\n\n')
bply(barplot(log=log, x='name', y='percent', fill='Name', filllab='Sample'),
        xlab='Fastqscreen Filtering steps', ylab='Percent')
cat('\n\n### Fastqscreen files: {.tabset} \n')
cat('\n\n#### Logs\n\n')
catlink(logs[['fastqscreen']], 'all', 'log')
for (sample in names(config[['sample']])) {
    cat(paste0('\n\n#### ', sample, '\n\n'))
    catlink(sample.list[[sample]][['screen']], sample, 'screen')
    catlink(sample.list[[sample]][['screen_SN']], sample, 'Serial Number')
    catlink(sample.list[[sample]][['screen_ids']], sample, 'fastq IDs vs. HTtools IDs')
    catlink(sample.list[[sample]][['screen_PBS']], sample, 'PBS-matching screen')
    catlink(sample.list[[sample]][['screen_PBS_SN']], sample, 'PBS-matching Serial Number')
    catlink(sample.list[[sample]][['screen_PBS_ids']], sample, 'PBS-matching fastq IDs vs. HTtools IDs')
}
```

## Blast

```{r, results='asis', eval=eval.opt}
cat('Blast files:\n')
for (sample in names(config[['sample']])) {
    catlink(logs[['blast']][[sample]], sample, 'log')
    catlink(sample.list[[sample]][['blast']], sample, 'blast')
}
```

## Filterblast

```{r, results='asis', eval=eval.opt}
logheader <- c('total_seq', 'lines', 'start_over1', 'high_evalue')
log <- read.csv(logs[['integration']], sep='\t', header=TRUE, row.names=1)
knitr::kable(log[logheader, , drop=FALSE])

newcols <- c(NA, 'total', NA)
for (chro in sort(chro_list)) {
    newcols <- c(newcols, chro, NA)
}
log_per_chro <- list()
for (sample in names(log)) {
    log_per_chro[[sample]]  <- log %>%
        mutate(feature = rownames(.), sample = sample) %>%
        filter(!feature %in% logheader) %>%
        select(feature, !!sample, sample) %>%
        separate(!!sample, into=newcols, sep=',|:|\\}') %>%
        mutate_at(vars(-feature, -sample), as.numeric) %>%
        pivot_longer(names(.)[!names(.) %in% c('feature', 'sample')])

    log_per_chro[[sample]] <- log_per_chro[[sample]] %>%
        dplyr::group_by(feature) %>%
        dplyr::mutate(percent = value / (sum(value, na.rm=TRUE)/2))
}

log_per_chro <- bind_rows(log_per_chro)
log_per_chro$feature <- factor(log_per_chro$feature, levels=rev(sort(unique(log_per_chro$feature))))
cat('\n\n### Raw reads\n\n')
bp <- barplot(log=log_per_chro, x='name', fill='sample', filllab='Sample') +
    facet_wrap(~feature, nrow=2)
bply(bp, xlab='Chromosome')
cat('\n\n### Percent reads\n\n')
bp <- barplot(log=log_per_chro, x='name', y='percent', fill='sample', filllab='Sample') +
    facet_wrap(~feature, nrow=2)
bply(bp, xlab='Chromosome', ylab='Percent')
cat('### Filterblast files: {.tabset} \n')
cat('\n\n#### Logs\n\n')
catlink(logs[['integration']], 'all', 'log')
for (sample in names(config[['sample']])) {
    cat(paste0('\n\n#### ', sample, '\n\n'))
    catlink(sample.list[[sample]][['preintegration']], sample, 'uniquely mapped integration (contains homologous recombination positions)')
    catlink(sample.list[[sample]][['preintegration_multimatch']], sample, 'multimatch integration (contains homologous recombination positions)')
}
cat('\n\nSee next filtering step for number of integrations excluding homologous recombination positions.\n\n')

```

## Homologous recombination

```{r, results='asis', eval=eval.opt}
log <- read.csv(logs[['location']], sep='\t', header=TRUE, row.names=1)
hrlog <- log[grep('integration', rownames(log)), ] %>%
    mutate(feature = rownames(.)) %>%
    pivot_longer(names(.)[!names(.) %in% c('feature')]) %>%
    separate(name, into=c('sample', 'type')) %>%
    mutate(type = factor(type, levels = c('SSP', 'indpt', 'dupl')))
hrlog <- hrlog %>%
    dplyr::group_by(sample, type) %>%
    dplyr::mutate(percent = value / (sum(value)/2))
cat('\n\n### Raw reads\n\n')
bp <- barplot(log=hrlog, x='feature', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Feature')
cat('\n\n### Percent reads\n\n')
bp <- barplot(log=hrlog, x='feature', y='percent', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Feature', ylab='Percent')
cat('\n\n### Homologous recombination files: {.tabset} \n')
cat('\n\n#### Logs\n\n')
catlink(logs[['location']], 'all', 'log')
for (sample in names(config[['sample']])) {
    cat(paste0('\n\n#### ', sample, '\n\n'))
    catlink(sample.list[[sample]][['ltr5']], sample, "homologous recombination at preexisting Tf2 5' LTRs")
    catlink(sample.list[[sample]][['ltr3']], sample, "homologous recombination at preexisting Tf2 3' LTRs")
    catlink(sample.list[[sample]][['sololtr']], sample, "homologous recombination at preexisting solo-LTRs")
}
```

```{r exclude_pos, results='asis', eval=eval.opt}
if (config[['exclude']] != 'na') {
    cat('\n\n## Excluded positions\n\n')
    exlog <- log[c('unfiltered_integration', 'true_integration', 'excluded_position'),] %>%
        mutate(feature = rownames(.)) %>%
        pivot_longer(names(.)[!names(.) %in% c('feature')]) %>%
        separate(name, into=c('sample', 'type')) %>%
        mutate(type = factor(type, levels = c('SSP', 'indpt', 'dupl')))
    exlog <- exlog %>%
        dplyr::group_by(sample, type) %>%
        dplyr::mutate(percent = value / (sum(value)/2))
    cat('\n\n### Raw reads\n\n')
    bp <- barplot(log=exlog, x='feature', fill='sample', filllab='Sample') +
        facet_wrap(~type, nrow=1)
    subchunkify(bply(bp, xlab='Feature'))
    cat('\n\n### Percent reads\n\n')
    bp <- barplot(log=exlog, x='feature', y='percent', fill='sample', filllab='Sample') +
        facet_wrap(~type, nrow=1)
    subchunkify(bply(bp, xlab='Feature', ylab='Percent'))
    cat('\n\n### Excluded position(s) files: {.tabset} \n')
    cat('\n\n#### Logs\n\n')
    catlink(logs[['location']], 'all', 'log')
    for (sample in names(config[['sample']])) {
        cat(paste0('\n\n#### ', sample, '\n\n'))
        catlink(sample.list[[sample]][['exclude']], sample, "excluded positions")
    }
}
```

# Results

## Integration details {.tabset}

```{r, results='asis', eval=!eval.opt}
cat('\n\nDisabled when location file provided in config\n\n')
```

```{r, results='asis'}
plot.heatmap <- function(sample.list, namecol, loc, intype){
    df <- list()
    for (sample in names(sample.list)) {
        df[[sample]] <- read.csv(sample.list[[sample]][[loc]], sep='\t')
        df[[sample]] <- df[[sample]] %>%
            mutate(SSP = ifelse('SSP' %in% names(df[[sample]]), SSP, 1)) %>%
            select(names(.)[names(.) %in% c(namecol, intype)]) %>%
            rename_at(vars(-!!namecol), paste0, '_', sample)
    }
    df <- df %>%
        reduce(full_join, by=namecol) %>%
        select(-!!namecol)
    df[is.na(df)] <- 0

    sampleDists <- cor(df)
    sampleDists[is.na(sampleDists)] <- 0
    sampleDistMatrix <- as.matrix(sampleDists)
    colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
    if (length(names(sample.list)) > 1) {
    heatmaply(sampleDistMatrix,
              scale='none',
              col=colors,
              cellnote=sampleDistMatrix,
              showticklabels=c(FALSE,TRUE))
    } else {
        cat('Not enough samples to plot heatmap')
    }
}
```


```{r, results='asis', eval=eval.opt}
cat('\n\n### Total\n\n')
tlog <- log[grepl('total', rownames(log)), ]
tlog['total',] <- colSums(tlog)
tlog <- tlog %>%
    mutate(feature = rownames(.)) %>%
    pivot_longer(names(.)[!names(.) %in% c('feature')]) %>%
    separate(name, into=c('sample', 'type')) %>%
    mutate(type = factor(type, levels = c('SSP', 'indpt', 'dupl')))
tlog <- tlog %>%
    dplyr::group_by(sample, type) %>%
    dplyr::mutate(percent = value / (sum(value)/2))
cat('\n\n#### Raw reads\n\n')
bp <- barplot(log=tlog, x='feature', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Feature')
cat('\n\n#### Percent reads\n\n')
bp <- barplot(log=tlog, x='feature', y='percent', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Feature', ylab='Percent')
cat('\n\n#### Correlation heatmap {.tabset}\n\n')
for (intype in unique(tlog[['type']])) {
    cat(paste0('\n\n##### ', intype, '\n\n'))
    subchunkify(plot.heatmap(sample.list, 'id', 'location', intype))
}

cat('\n\n### Intergenic \n\n')
intlog <- log[grepl('intergenic', rownames(log)), ] %>%
    mutate(feature = rownames(.)) %>%
    pivot_longer(names(.)[!names(.) %in% c('feature')]) %>%
    separate(name, into=c('sample', 'type'))
intlog <- intlog %>%
    mutate(feature = factor(feature, levels = unique(intlog$feature)),
    type = factor(type, levels = c('SSP', 'indpt', 'dupl')))
intlog <- intlog %>%
    dplyr::group_by(sample, type) %>%
    dplyr::mutate(percent = value / (sum(value)/3))
cat('\n\n#### Raw reads\n\n')
bp <- barplot(log=intlog, x='feature', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Intergenic Feature')
cat('\n\n#### Percent reads\n\n')
bp <- barplot(log=intlog, x='feature', y='percent', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='Intergenic Feature', ylab='Percent')
cat('\n\n#### Correlation heatmap {.tabset}\n\n')
for (intype in unique(tlog[['type']])) {
    cat(paste0('\n\n##### ', intype, '\n\n'))
    subchunkify(plot.heatmap(sample.list, 'name', 'intergenic', intype))
}

cat('\n\n### ORF \n\n')
orflog <- log[grepl('ORF', rownames(log)), ] %>%
    mutate(feature = rownames(.)) %>%
    pivot_longer(names(.)[!names(.) %in% c('feature')]) %>%
    separate(name, into=c('sample', 'type'))
orflog <- orflog %>%
    mutate(feature = factor(feature, levels = unique(orflog$feature)), 
    type = factor(type, levels = c('SSP', 'indpt', 'dupl')))
orflog <- orflog %>%
    dplyr::group_by(sample, type) %>%
    dplyr::mutate(percent = value / (sum(value)/2))
cat('\n\n#### Raw reads\n\n')
bp <- barplot(log=orflog, x='feature', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='ORF Feature')
cat('\n\n#### Percent reads\n\n')
bp <- barplot(log=orflog, x='feature', y='percent', fill='sample', filllab='Sample') +
    facet_wrap(~type, nrow=1)
bply(bp, xlab='ORF Feature', ylab='Percent')
cat('\n\n#### Correlation heatmap {.tabset}\n\n')
for (intype in unique(tlog[['type']])) {
    cat(paste0('\n\n##### ', intype, '\n\n'))
    subchunkify(plot.heatmap(sample.list, 'name', 'ORF', intype))
}
```

```{r summary, results='asis', eval=eval.opt}
cat('\n\n## Integration summary\n')
datatable(log,
          rownames=TRUE,
          filter='bottom',
          options=list(autoWidth=TRUE,
                       pageLength = 15)
          )
cat('\n\n## Integration and location files {.tabset} \n')
cat('\n\n### Logs\n\n')
catlink(logs[['location']], 'all', 'log')
for (sample in names(config[['sample']])) {
    cat(paste0('\n\n### ', sample, '\n\n'))
    catlink(sample.list[[sample]][['integration']], sample, 'integration (homologous recombination excluded)')
    catlink(sample.list[[sample]][['location']], sample, "location")
    catlink(sample.list[[sample]][['intergenic']], sample, "intergenic")
    catlink(sample.list[[sample]][['ORF']], sample, "ORF")
}
```

## ORFmap

```{r, results='asis'}
count_interval <- function(location, section, start, end) {
    # subset location for the section and interval start, end
    subloc <- location %>%
        filter(location %in% section,
               distance >= start,
               distance < end)
    n <- list()
    n[['SSP']] <- dim(subloc)[1]
    n[['indpt']] <- sum(subloc[['indpt']])
    n[['dupl']] <- sum(subloc[['dupl']])
    return(n)
}
```

```{r, results='asis', fig.width=12, fig.height=6}
orfmap <- list()
totallog <- NULL

for (sample in names(sample.list)) {
    location <- read.csv(sample.list[[sample]][['location']], sep='\t', header=TRUE)
    # filter out chromosomes not in config[['short_chro_list']]
    location <- location %>%
        filter(chr %in% config[['short_chro_list']])
    rownames(location) <- location[['id']]
    #calculate totals
    totallog <- bind_rows(totallog, tibble(sample = c(sample, sample, sample),
                               type = c('SSP', 'indpt', 'dupl'),
                               nb = c(dim(location)[1], sum(location$indpt), sum(location$dupl))
                               ))
    # upstream
    for (pos in seq(config[['orf_map_window']]-1, 1, -config[['orf_map_interval']])) {
        name <- paste0('upstream_', pos-config[['orf_map_interval']] +1, '_', pos)
        orfmap[[sample]][[name]] <- count_interval(location,
                                                            'UP',
                                                            pos-config[['orf_map_interval']] +1,
                                                            pos+1)
        orfmap[[sample]][[name]] <- bind_rows(orfmap[[sample]][[name]])
    }
    # in ORF
    location_frac <- location %>%
        mutate(distance = ifelse(upstrand == '+',
                                 (position-upstart) / (upend-upstart),
                                 (upend-position) / (upend-upstart)))
    n_orf <- config[['avg_orf_length']] / config[['orf_map_interval']]
    for (pos in seq(1, n_orf, 1)) {
        name <- paste0('in_ORF_', pos)
        orfmap[[sample]][[name]] <- count_interval(location_frac,
                                                            c('IN_start', 'IN_end'),
                                                            (pos-1)/n_orf,
                                                            pos/n_orf)
        orfmap[[sample]][[name]] <- bind_rows(orfmap[[sample]][[name]]) 
    }
    # downstream
    for (pos in seq(0, config[['orf_map_window']]-1, config[['orf_map_interval']])) {
        name <- paste0('downstream_', pos, '_', pos+config[['orf_map_interval']]-1)
        orfmap[[sample]][[name]] <- count_interval(location,
                                                            'DOWN',
                                                            pos,
                                                            pos+config[['orf_map_interval']])
        orfmap[[sample]][[name]] <- bind_rows(orfmap[[sample]][[name]])
    }
    orfmap[[sample]] <- bind_rows(orfmap[[sample]], .id="Interval")
}
orfmap <- bind_rows(orfmap, .id='Sample')
orfmaplong <- orfmap %>%
    pivot_longer(c('SSP', 'indpt', 'dupl')) %>%
    mutate(Interval = factor(Interval, levels=unique(orfmap[['Interval']])))
orfmaplong <- orfmaplong %>%
    mutate(name = factor(name, levels=unique(orfmaplong[['name']])),
           to.join=paste0(Sample, '_', name))
# original perl script is miscounting the number of SSP for non SN, adding 1
# check if SN by checking if position is numeric
if ((config[['legacy_mode']] == TRUE) & !(is.numeric(config[['sample']][[sample]][['SN_position']]))) {
    totallog <- totallog %>%
        mutate(nb = ifelse(type == 'SSP', nb +1, nb))
}
totallog <- totallog %>%
    mutate(to.join = paste0(sample, '_', type))
orfmaplong <- merge(orfmaplong, totallog %>% select(to.join, nb), by = 'to.join', all.x=T)
orfmaplong <- orfmaplong %>%
    dplyr::group_by(name, Sample) %>%
    dplyr::mutate(percent = value * 100 / nb) %>%
    ungroup() %>%
    arrange(name)

percentmap <- orfmaplong %>%
    select(-to.join) %>%
    arrange(Sample, Interval) %>%
    pivot_wider(names_from=name, values_from=c(value, percent, nb))
fn <- paste0(mydir, '/ORFmap/ORFmap.tsv')

write.table(percentmap, file=fn, sep='\t', quote=FALSE, row.names=FALSE)
catlink(fn, 'all', 'ORFmap table')

cat('\n\n### Raw reads\n\n')
bp <- barplot(log=orfmaplong, x='Interval', fill='Sample', filllab='Sample') +
    facet_wrap(~name, nrow=3) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
bply(bp, xlab='Distance to ORF')
fn <- paste0(mydir, '/ORFmap/ORFmap_rawreads.pdf')
ggsave(fn)
catlink(fn, 'all', 'ORFmap in raw reads PDF')
cat('\n\n### Percent reads\n\n')
bp <- barplot(log=orfmaplong, x='Interval', y='percent', fill='Sample', filllab='Sample') +
    facet_wrap(~name, nrow=3) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
bply(bp, xlab='Distance to ORF', ylab='Percent')
fn <- paste0(mydir, '/ORFmap/ORFmap_percentreads.pdf')
ggsave(fn)
catlink(fn, 'all', 'ORFmap in percentages PDF')
```



## Heatmap of ORF and intergenic

Counts are normalized relative to the total number of true integrations in their respective categories (SSP, indpt is applicable, dupl).
Regions are sorted by decreasing number of duplicates and the top 50 are plotted.

```{r intergenic, results='asis'}
inter <- list()
regions <- list()
totallog <- ungroup(totallog)
for (sample in names(sample.list)) {
    inter[[sample]] <- read.csv(sample.list[[sample]][['intergenic']], sep='\t', header=TRUE)
    inter[[sample]] <- inter[[sample]] %>%
        mutate(id = paste(name, name_dn, sep='_'),
               norm_SSP = SSP / as.numeric(totallog %>% filter(sample==!!sample, type=='SSP') %>% select(nb)),
               norm_dupl = dupl / as.numeric(totallog %>% filter(sample==!!sample, type=='dupl') %>% select(nb))
        )
    if ('indpt' %in% names(inter[[sample]])) {
        inter[[sample]] <- inter[[sample]] %>%
            mutate(norm_indpt = indpt / as.numeric(totallog %>% filter(sample==!!sample, type=='indpt') %>% select(nb)))
    }
    row.names(inter[[sample]]) <- inter[[sample]][['id']]
    names(inter[[sample]]) <- paste(sample, names(inter[[sample]]), sep = "_")
}
inter <- bind_cols(inter)
select_cols <- c(names(inter)[grepl('id', names(inter))][1],
                 #names(inter)[grepl('chrom', names(inter))],
                 names(inter)[grepl('SSP', names(inter))],
                 names(inter)[grepl('indpt', names(inter))],
                 names(inter)[grepl('dupl', names(inter))])
regions[['intergenic']] <- inter[select_cols]
colnames(regions[['intergenic']])[1] <- 'id'

# write to output file
fn <- paste0(mydir, '/heatmap/heatmap_intergenic.tsv')
write.table(regions[['intergenic']], file=fn, sep='\t', quote=FALSE)
catlink(fn, 'all', 'heatmap intergenic table')
```

```{r orf, results='asis'}
orf <- list()
for (sample in names(sample.list)) {
    orf[[sample]] <- read.csv(sample.list[[sample]][['ORF']], sep='\t', header=TRUE)
    orf[[sample]] <- orf[[sample]] %>%
        # SPAC1556.06 ORF name is duplicated, set id to name_position
        mutate(id = paste0(name, '_', start),
               norm_SSP = SSP / as.numeric(totallog %>% filter(sample==!!sample, type=='SSP') %>% select(nb)),
               norm_dupl = dupl / as.numeric(totallog %>% filter(sample==!!sample, type=='dupl') %>% select(nb))
        )
    if ('indpt' %in% names(orf[[sample]])) {
        orf[[sample]] <- orf[[sample]] %>%
            mutate(norm_indpt = indpt / as.numeric(totallog %>% filter(sample==!!sample, type=='indpt') %>% select(nb)))
    }
    row.names(orf[[sample]]) <- orf[[sample]][['id']]
    names(orf[[sample]]) <- paste(sample, names(orf[[sample]]), sep = "_")
}
orf <- bind_cols(orf)
select_cols <- c(names(orf)[grepl('id', names(orf))][1],
                 #names(orf)[grepl('chrom', names(orf))],
                 names(orf)[grepl('SSP', names(orf))],
                 names(orf)[grepl('indpt', names(orf))],
                 names(orf)[grepl('dupl', names(orf))])
regions[['orf']] <- orf[select_cols]
colnames(regions[['orf']])[1] <- 'id'

# write to output file
fn <- paste0(mydir, '/heatmap/heatmap_orf.tsv')
write.table(regions[['orf']], file=fn, sep='\t', quote=FALSE)
catlink(fn, 'all', 'heatmap ORF table')
```


```{r heatmap, results='asis'}
# sort by highest in each of the sample, and plot heatmapply for top 200, for intergenic and for ORF
for (region in c('intergenic', 'orf')) {
    cat(paste0('\n\n### ', region, ' {.tabset}\n\n'))
    for (intype in c('SSP', 'indpt', 'dupl')) {
        cat(paste0('\n\n#### ', intype, ' {.tabset}\n\n'))
        for (sample in names(sample.list)) {
            cat(paste0('\n\n##### Sorted by ', sample, '\n\n'))
            sort_col <- paste0(sample, '_norm_', intype)
            if (sort_col %in% names(regions[[region]])) {
                sub <- regions[[region]] %>%
                    arrange(desc(!!as.name(sort_col))) %>%
                    tibble::column_to_rownames('id')
                select_cols <- c(names(sub)[grepl(paste0('_norm_', intype), names(sub))])
                subchunkify(heatmaply(sub[1:50,select_cols, drop=FALSE],
                              xlab = "Samples",
                              Rowv = FALSE,
                              Colv = FALSE))
            } else { cat('\n\nNo independent integration\n\n')}
        }
    }
}
```


# Session info

```{r}
sessionInfo()
```

